src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -461,6 +472,138 @@ bool ExpressionAnalyzer::makeAggregateDescriptions(ActionsDAGPtr & actions)
+        window_description.full_sort_description.insert(
+            window_description.full_sort_description.end(),
+            window_description.order_by.begin(),
+            window_description.order_by.end());
+
+        WindowFunctionDescription window_function;
+        window_function.window_name = window_description.window_name;
+        window_function.function_node = function_node;
+        window_function.column_name
+            = window_function.function_node->getColumnName();
+        window_function.function_parameters
+            = window_function.function_node->parameters
+                ? getAggregateFunctionParametersArray(
+                    window_function.function_node->parameters)
+                : Array();
+
+        // Requiring a constant reference to a shared pointer to non-const AST
+        // doesn't really look sane, but the visitor does indeed require it.
+        // Hence we clone the node (not very sane either, I know).
+        getRootActionsNoMakeSet(window_function.function_node->clone(),
+            true, actions);
+
+        const ASTs & arguments
+            = window_function.function_node->arguments->children;
+        window_function.argument_types.resize(arguments.size());
+        window_function.argument_names.resize(arguments.size());
+        const auto & index = actions->getIndex();
+        for (size_t i = 0; i < arguments.size(); ++i)
+        {
+            const std::string & name = arguments[i]->getColumnName();
+
+            auto it = index.find(name);
+            if (it == index.end())
+            {
+                throw Exception(ErrorCodes::UNKNOWN_IDENTIFIER,
+                    "Unknown identifier '{}' in window function '{}'",
+                    name, window_function.function_node->formatForErrorMessage());
+            }
+
+            window_function.argument_types[i] = (*it)->result_type;
+            window_function.argument_names[i] = name;
+        }
+
+        AggregateFunctionProperties properties;
+        window_function.aggregate_function
+            = AggregateFunctionFactory::instance().get(
+                window_function.function_node->name,
+                window_function.argument_types,
+                window_function.function_parameters, properties);
+
+        window_function_descriptions.push_back(window_function);
+
+        if (auto it = window_descriptions.find(window_description.window_name);
+            it != window_descriptions.end())
+        {
+            assert(it->second.full_sort_description
+                == window_description.full_sort_description);
+        }
+        else
+        {
+            window_descriptions.insert({window_description.window_name,
+                window_description});
+        }
+    }
+
+    // Populate the reverse map.
+    for (const auto & f : window_function_descriptions)
+    {
+        window_descriptions[f.window_name].window_functions.push_back(f);
```

Comment:
So, we use `window_name ` as grouping key, so it is equal for each window function from one window. OK.
It is interesting if `window_name` is used somewhere else. If not, we can remove this field form `WindowFunctionDescription`

Reply:
It turned out to be redundant after we group the functions by windows, yes. But I'm not sure where else to put it... Maybe we can just group right away, and remove the common array of window functions and this separate grouping step.

Reply:
I managed to remove it.

Reply:
This also addresses a  couple of the above comments about maps and duplication of `window_name`.